reconcile/utils/saasherder.py

```diff
@@ -1441,6 +1505,93 @@ def get_upstream_jobs_diff_saas_file(
 
         return trigger_specs
 
+    def get_container_images_diff(
+        self, dry_run: bool
+    ) -> list[TriggerSpecContainerImage]:
+        results = threaded.run(
+            self.get_container_images_diff_saas_file,
+            self.saas_files,
+            self.thread_pool_size,
+            dry_run=dry_run,
+        )
+        return list(itertools.chain.from_iterable(results))
+
+    def get_container_images_diff_saas_file(
+        self, saas_file: dict[str, Any], dry_run: bool
+    ) -> list[TriggerSpecContainerImage]:
+        saas_file_name = saas_file["name"]
+        timeout = saas_file.get("timeout") or None
+        pipelines_provider = self._get_pipelines_provider(saas_file)
+        github = self._initiate_github(saas_file)
+        trigger_specs: list[TriggerSpecContainerImage] = []
+        for rt in saas_file["resourceTemplates"]:
+            rt_name = rt["name"]
+            url = rt["url"]
+            for target in rt["targets"]:
+                try:
+                    image = target.get("image")
+                    if not image:
+                        continue
+                    ref = target["ref"]
+                    hash_length = rt.get("hash_length") or self.settings["hashLength"]
+                    get_commit_sha_options = {
+                        "url": url,
+                        "ref": ref,
+                        "github": github,
+                        "hash_length": hash_length,
+                    }
+                    desired_image_tag = self._get_commit_sha(get_commit_sha_options)
+                    # don't trigger if image doesn't exist
+                    image_registry = f"{image['org']['instance']['url']}/{image['org']['name']}/{image['name']}"
+                    image_uri = f"{image_registry}:{desired_image_tag}"
+                    image_patterns = saas_file["imagePatterns"]
+                    image_auth = self._initiate_image_auth(saas_file)
+                    error_prefix = f"[{saas_file_name}/{rt_name}] {ref}:"
+                    error = self._check_image(
+                        image_uri, image_patterns, image_auth, error_prefix
+                    )
```

Comment:
What happens if there are two pushes close enough that build for the first is not ended? If I understand correctly, this will continue until the second one finishes to be built...

Reply:
yup, that makes sense. this is consistent with the behavior of `openshift-saas-deploy-trigger-upstream-jobs`.

for the sake of the discussion, let's define the `intermediate` image as the image created by the first push, and the `final` image as the one created by the second (quick) push.

the alternative to this behavior is to:
- wait for the intermediate image to be created
- trigger a deployment
- wait for the final image to be created
- trigger a deployment

depending on the implementation of the entity building the image, there may be situations where builds will be cancelled (common in jenkins) and the intermediate image (the first push before the quick second) will not be created.

on the other hand, the final image will always be created. so i feel this behavior makes sense.

one possible concern is that a team may promote to production the intermediate image, which wasn't deployed, and hence wasn't tested.
in case the image wasn't created as described above we are covered, because the promotion MR checks that the image exists.
in case the image was created, a promotion may be feasible, but we offer and recommend teams to use gated promotions. a tool can not always provide guard rails and automation can not always cover bad practices.

/braindump

Reply:
When I started thinking abut this, I had in my mind in a case in which the intermediate had a migration. I couldn't think of a case in which the behaviour could be problematic (yours is also interesting).

The migration case should be covered because migration code is always executed, usually with no effect as the state of the DB is the state in the code. We don't really care if it is executed in the intermediate or the last image, honestly.

Yours is definitely not covered, but it would be honestly weird.

If this is consistent with the other promotion code, let's keep it. I would only document it somehow, since I only learnt about this by reading a few times the code 😅